nannyml/data_quality/range/calculator.py

```diff
@@ -0,0 +1,252 @@
+                    raise InvalidArgumentsException(
+                        f"column_names elements should be either a column name string or a list of strings, found\n{el}"
+                    )
+            self.column_names = column_names
+        else:
+            raise InvalidArgumentsException(
+                "column_names should be either a column name string or a list of columns names strings, "
+                "found\n{column_names}"
+            )
+        self.result: Optional[Result] = None
+
+        # threshold strategy is the same across all columns
+        self.threshold = threshold
+        self._upper_alert_thresholds: Dict[str, Optional[float]] = {column_name: 0 for column_name in self.column_names}
+        self._lower_alert_thresholds: Dict[str, Optional[float]] = {column_name: 0 for column_name in self.column_names}
+
+        self.lower_threshold_value_limit: float = 0
+        self.upper_threshold_value_limit: Optional[float] = None
+        self.normalize = normalize
+
+        if self.normalize:
+            self.data_quality_metric = 'out_of_range_values_rate'
+            self.upper_threshold_value_limit = 1
+        else:
+            self.data_quality_metric = 'out_of_range_values_count'
+            self.upper_threshold_value_limit = np.nan
+
+        #object tracks values as list [min,max]
+        self._continuous_val_ranges: Dict[str, list] = {column_name: list() for column_name in self.column_names}
+
+
+    def _calculate_out_of_range_stats(self, data: pd.Series, lower_bound: float, upper_bound: float):
+        # to do make this calc out of range stats
+        count_tot = data.shape[0]
+        count_out_of_range  = ((data < lower_bound) | (data > upper_bound)).sum()
+        if self.normalize:
+            count_out_of_range = count_out_of_range / count_tot
+        return count_out_of_range
+
+    @log_usage(UsageEvent.DQ_CALC_VALUES_OUT_OF_RANGE_FIT, metadata_from_self=['normalize'])
+    def _fit(self, reference_data: pd.DataFrame, *args, **kwargs):
+        """Fits the drift calculator to a set of reference data."""
+        if reference_data.empty:
+            raise InvalidArgumentsException('data contains no rows. Please provide a valid data set.')
+
+        _list_missing(self.column_names, reference_data)
+
+        # All provided columns must be continuous
+        # We do not make int categorical
+        continuous_column_names, categorical_column_names = _split_features_by_type(reference_data, self.column_names)
+        if not set(self.column_names) == set(continuous_column_names):
+            raise InvalidArgumentsException(
+                f"Specified columns_names for NumericalRangeCalculator must all be continuous.\n"
+                f"Categorical columns found:\n{categorical_column_names}"
+            )
+        
+        for col in self.column_names:
+            self._continuous_val_ranges[col] = [reference_data[col].min(), reference_data[col].max()]
+
+        return self
```

Comment:
Missing call to `self._calculate` here! This will ensure we have results for the metric in the reference period too. Agreed: not that important here, as it is by default always 0, but this is how our little pattern works.